### PR TITLE
Extract CM into asm4s

### PR DESCRIPTION
**Pull request by danking**
Now the CM carries around an opaque `AnyRef` state value. It probably
shouldn't even carry this around, but this works for now. The extra
state value will be completely eliminated anyway in a follow up pull
request.

All references to `Type` in the CM were converted to `ClassTag`, which
is the only information we actually need.

### CODE REVIEW

**Comment by danking** [comment]
@cseed step 1 of compiler plan

File: src/main/scala/is/hail/asm4s/CM.scala [hunk 1]
```diff
@@ -1,20 +1,22 @@
...
 object preCM {
-  case class E(m: Map[String, (Type, Code[AnyRef])], fb: Function2Builder[Array[AnyRef],mutable.ArrayBuffer[AnyRef],AnyRef])
-  case class S(fa: Array[AnyRef], ec: EvalContext)
+  case class E(m: Map[String, (ClassTag[AnyRef], Code[AnyRef])], fb: Function2Builder[Array[AnyRef],mutable.ArrayBuffer[AnyRef],AnyRef])
+  case class S(fa: Array[AnyRef], st: AnyRef)
 
-  def emptyE = E(Map[String, (Type, Code[AnyRef])](), new Function2Builder[Array[AnyRef],mutable.ArrayBuffer[AnyRef],AnyRef]())
-  def emptyS(ec: EvalContext) = S(Array[AnyRef](), ec)
+  def emptyE = E(Map[String, (ClassTag[AnyRef], Code[AnyRef])](), new Function2Builder[Array[AnyRef],mutable.ArrayBuffer[AnyRef],AnyRef]())
+  def emptyS(ec: AnyRef) = S(Array[AnyRef](), ec)
```
Comment (cseed) [inline src/main/scala/is/hail/asm4s/CM.scala]: I think this needs more radical refactoring.

1. asm4s doens't need the extra binding mechanism for tracking Hail variables, it should support the for-based lexical scoping we talked about in the vectorized numeric operations example.

2.  I'd like to keep asm4s typed unless we're adding an UntypedCode.  I think it is possible with the above constraints.  That mean all this AnyRef nonsense has to go (again, an artifact of the expr language).

3. This should support general function builders.   I'm not 100% sure how to handle typing there.  Maybe FunctionBuilder should be parameterized by the return type of result().

File: src/main/scala/is/hail/asm4s/CMHelp.scala [hunk 2]
```diff
@@ -9,11 +9,11 @@ import scala.collection.mutable
 import scala.reflect.ClassTag
 
 // Without this, we can't use invokeStatic with CompilationHelp as a type.
-case class CompilationHelp() {}
+case class CMHelp() {}
```
Comment (cseed) [inline src/main/scala/is/hail/asm4s/CMHelp.scala]: I don't understand this.  Non-companion object functions aren't static?  By CompilationHelp, do you mean CMHelp?

File: src/main/scala/is/hail/expr/AST.scala [hunk 3]
```diff
@@ -236,7 +236,7 @@ sealed abstract class AST(pos: Position, subexprs: Array[AST] = Array.empty) {
   def runAggregator(ec: EvalContext): CPS[Any] = {
     val typedNames = ec.st.toSeq
       .sortBy { case (_, (i, _)) => i }
-      .map { case (name, (_, typ)) => (name, typ) }
+      .map { case (name, (_, typ)) => (name, typ.scalaClassTag.asInstanceOf[ClassTag[AnyRef]]) }
```
Comment (cseed) [inline src/main/scala/is/hail/expr/AST.scala]: These casts are tedious.  Let's see how it looks when you address the above comments.  I think some C[Any] and C[AnyRef] in the code should actually be C[_].

File: src/main/scala/is/hail/asm4s/CM.scala (hunk 1 above)
Comment (cseed) [inline src/main/scala/is/hail/asm4s/CM.scala]: It might be useful to rewrite/add a asm4s test that uses the for binding style.

File: src/main/scala/is/hail/asm4s/CMHelp.scala (hunk 2 above)
Comment (danking) [inline src/main/scala/is/hail/asm4s/CMHelp.scala]: Yeah, I should have updated the comment too.

Without this line, I cannot refer to `CMHelp` in a type context, in Scala. For example, I couldn't use `CMHelp` as the first type parameter in this expression:
```
Code.invokeStatic[CMHelp, Array[T], mutable.WrappedArray[T]]("__arrayToWrappedArray", a)
```

File: src/main/scala/is/hail/expr/AST.scala (hunk 3 above)
Comment (danking) [inline src/main/scala/is/hail/expr/AST.scala]: If we have an environment of the form `Map[Name, F[T]]` we cannot choose `T` to be `_` because, in Scala, the existential is bound at the edges of the type expression:
```
∃T. Map[Name, F[T]]
```
It is impossible to add key-value pairs to this map because the consumer doesn't know what `T` is.

What we want is:

```
Map[Name, ∃T. F[T]]
```

I think the only way to write this in Scala is:

```
trait Existential {
  type T
  def thing: T
}

def addElement(m: Map[String, Existential]): Map[String, Existential] =
  m + ("foo" -> new Existential() { val thing = 3 })
```

This seemed like a whole lot of nonsense for not much value.

Comment (danking) [inline src/main/scala/is/hail/expr/AST.scala]: Agreed though, very tedious. I thought about having `unsafeAnyRefScalaClassTag`

File: src/main/scala/is/hail/asm4s/CM.scala (hunk 1 above)
Comment (danking) [inline src/main/scala/is/hail/asm4s/CM.scala]: Correct: if we use Scala's binding structure, we will maintain type information (because we don't traffic our types into and out of a `Map` mediating their identity through `String`s).

I planned to move the environment off of `CM` and into the AST `compile` methods as a separate PR, but I can do it as a part of this one too, if you prefer that.

File: src/main/scala/is/hail/expr/AST.scala (hunk 3 above)
Comment (danking) [inline src/main/scala/is/hail/expr/AST.scala]: I suppose you really want either:

```scala
trait BoundExistential[U] {
  type T <: U
  def thing: T
}
def addElement(m: Map[String, BoundExistential[Number]]): Map[String, BoundExistential[Number]] =
  m + ("foo" -> new BoundExistential[Number]() { type T = Integer; val thing = 3 })
```

```scala
import scala.language.higherKinds
trait ConstrainedExistential[U[_]] {
  type T
  def pf: U[T]
  def thing: T
}
def addElement(m: Map[String, ConstrainedExistential[Numeric]]): Map[String, ConstrainedExistential[Numeric]] =
  m + ("foo" -> new ConstrainedExistential[Numeric]() {
    type T = Int; val pf = implicitly[Numeric[Int]]; val thing = 3 })
```

File: src/main/scala/is/hail/asm4s/CM.scala (hunk 1 above)
Comment (cseed) [inline src/main/scala/is/hail/asm4s/CM.scala]: OK, great!  Let's keep pushing.

File: src/main/scala/is/hail/expr/AST.scala (hunk 3 above)
Comment (cseed) [inline src/main/scala/is/hail/expr/AST.scala]: Scala has `forSome` for existential types, so you don't need to cook up your own `Existential.

 > in Scala, the existential is bound at the edges of the type expression

Can we double-check this?

```
scala> val b: mutable.Map[String, Array[T] forSome { type T }] = mutable.Map("foo" -> Array(1, 2, 3))
b: scala.collection.mutable.Map[String,Array[_]] = Map(foo -> Array(1, 2, 3))
```

Note the printed type, vs:

```
scala> val c: mutable.Map[String, Array[T]] forSome { type T } = mutable.Map("foo" -> Array(1, 2, 3))
c: scala.collection.mutable.Map[String,Array[T]] forSome { type T } = Map(foo -> Array(1, 2, 3))
```

And they behave as expected:

```
scala> b += "bar" -> Array("baz")
res2: b.type = Map(foo -> Array(1, 2, 3), bar -> Array(baz))

scala> c += "bar" -> Array("baz")
<console>:14: error: type mismatch;
 found   : String("baz")
 required: T
       c += "bar" -> Array("baz")
```

Also:

```
scala> val d: mutable.Map[String, Array[_]] = mutable.Map("foo" -> Array(1, 2, 3))
d: scala.collection.mutable.Map[String,Array[_]] = Map(foo -> Array(1, 2, 3))

scala> d += ("bar" -> Array("baz"))
res5: d.type = Map(foo -> Array(1, 2, 3), bar -> Array(baz))
```

Comment (cseed) [inline src/main/scala/is/hail/expr/AST.scala]: I'm not totally clear how F[_] is used downstream so I might be missing the point.

Comment (danking) [inline src/main/scala/is/hail/expr/AST.scala]: Yes, you're absolutely correct, I somehow forgot how this worked.

Let me try again, I think some issue came up because I wanted this:

```scala
Map[String, (ClassTag[T], Code[T]) forSome { type T }]
```

**Comment by danking** [comment]
@cseed back to you, I inserted existential types everywhere. I managed this with only two `unsafe` functions: in let-binding and in lambda aggregators. In both cases, the type of the `Code[T]` should be correct if the compilation process is correct. Adding a `checkcast` would create a tremendous amount of unnecessary byte code. I suspect returning the type as a part of the compilation process would enable me to remove the let-binding `unsafe`. I'm not sure about the lambda aggregator case. I'm unconcerned since all that code is disappearing soon anyway.

**Comment by cseed** [comment]
I'll look at this, but you're going to put the environment stuff back in AST, right?

**Comment by danking** [comment]
@cseed I was going to make a separate PR for that, yes.

**Comment by danking** [comment]
Closing until #1740 lands.
